src/domain/subscription/subscription.service.ts

```diff
@@ -1,15 +1,76 @@
-import { Inject, Injectable } from '@nestjs/common';
+import { HttpStatus, Inject, Injectable } from '@nestjs/common';
 import { SUBSCRIPTION_REPOSITORY, SubscriptionRepository } from './subscription.repository';
 import { SubscriptionWithChannelResult } from './result/subscription-with-channel.result';
+import {
+  STREAMING_SERVER_CLIENT,
+  StreamingServerClient,
+} from '../streamer/client/streaming-server.client';
+import { STREAMER_REPOSITORY, StreamerRepository } from '../streamer/streamer.repository';
+import { DomainCustomException } from '../common/errors/domain-custom-exception';
+import { DomainErrorCode } from '../common/errors/domain-error-code';
+import { CACHE_SERVICE, CacheService } from '../common/cache/cache.service';
 
 @Injectable()
 export class SubscriptionService {
   constructor(
     @Inject(SUBSCRIPTION_REPOSITORY)
-    private readonly subscriptionRepository: SubscriptionRepository
+    private readonly subscriptionRepository: SubscriptionRepository,
+    @Inject(STREAMER_REPOSITORY)
+    private readonly streamerRepository: StreamerRepository,
+    @Inject(STREAMING_SERVER_CLIENT)
+    private readonly streamingServerClient: StreamingServerClient,
+    @Inject(CACHE_SERVICE)
+    private readonly cacheService: CacheService
   ) {}
 
-  async getSubscriptions(): Promise<SubscriptionWithChannelResult[]> {
-    return await this.subscriptionRepository.getSubscriptions();
+  private readonly IS_CHANGED_KEY = process.env.IS_CHANGED_KEY || 'monitoring:isChanged';
+
+  async getSubscriptions(userId: number): Promise<SubscriptionWithChannelResult[]> {
+    return await this.subscriptionRepository.getSubscriptions(userId);
+  }
+
+  async subscribe(userId: number, channelUrl: string): Promise<void> {
+    const subscriptions = await this.subscriptionRepository.getSubscriptions(userId);
+    const SUBSCRIPTION_LIMIT = Number(process.env.SUBSCRIPTION_LIMIT);
+
+    if (subscriptions.length === SUBSCRIPTION_LIMIT) {
+      throw new DomainCustomException(
+        HttpStatus.FORBIDDEN,
+        DomainErrorCode.SUBSCRIPTION_LIMIT_EXCEEDED
+      );
+    }
+
+    const channelInfo = await this.streamingServerClient.getChannelInfo(channelUrl);
+
+    const streamer = await this.streamerRepository.findStreamerByChannelId(channelInfo.channelId);
+
+    if (streamer) {
+      const subscription = await this.subscriptionRepository.findSubscription(
+        userId,
+        streamer.streamerId
+      );
+
+      if (subscription && subscription.isConnected) {
+        throw new DomainCustomException(
+          HttpStatus.CONFLICT,
+          DomainErrorCode.DUPLICATED_SUBSCRIPTION
+        );
+      }
```

Comment:
```suggestion
subscription.isConnected();
```
절차지향적인 코드 느낌이 많이 납니다.
도메인 중심적인 코드로 한번 리팩토링 해보면 어떨까요??

```javascript
export class SubscriptionEntity {
  constructor(
    public readonly subscriptionId: number,
    public readonly userId: number,
    public readonly streamerId: number,
    public readonly isConnected: boolean,
    public readonly createdAt: Date,
    public readonly updatedAt: Date
  ) {}

  isConnected(): void {
    if (this && this.isConnected) {
        throw new DomainCustomException(
          HttpStatus.CONFLICT,
          DomainErrorCode.DUPLICATED_SUBSCRIPTION
        );
      } 
  }
}
```

Reply:
- SubscriptionEntity
```ts
connect() {
    if (this.isConnected) {
      throw new DomainCustomException(HttpStatus.CONFLICT, DomainErrorCode.DUPLICATED_SUBSCRIPTION);
    }

    this.isConnected = true;

    return this;
  }
```

이런식으로 적용해봤습니다. 🙂